src/cache/mod.rs

```diff
@@ -72,34 +93,51 @@ where
     ///
     pub async fn set(&self, key: T, value: V, custom_duration: Option<Duration>) -> Option<V>
     where
-        T: Eq + Hash,
+        T: Eq + Hash + Clone,
     {
-        self.items
+        let replaced = self
+            .items
             .write()
             .await
             .insert(
                 key,
                 Item::new(value, custom_duration.or(self.item_duration)),
             )
-            .map(|item| item.object)
+            .map(|item| item.object);
+        self.remove_expired().await;
+        self.drop_excess().await;
+        replaced
     }
 
     ///
-    pub async fn remove_expired(&self)
-    where
-        T: Eq + Hash + Clone,
-    {
-        let expired_keys: Vec<T> = self
-            .items
-            .read()
-            .await
+    pub async fn remove_expired(&self) {
+        let read_items = self.items.read().await;
+        let expired_keys: Vec<_> = read_items
             .iter()
             .filter(|(_, item)| item.expired())
-            .map(|(key, _)| key.clone())
+            .map(|(key, _)| key)
             .collect();
 
         for key in expired_keys {
-            let _ = self.items.write().await.remove(&key);
+            let _ = self.items.write().await.remove(key);
+        }
+    }
+
+    /// removs keys beyond capacity
+    async fn drop_excess(&self) {
+        let len = self.len().await;
+        if len > self.capacity {
+            let excess = len - self.capacity;
+            let read_items = self.items.read().await;
+            let mut items = read_items.iter().collect_vec();
+
+            // sort
```

Comment:
no need to comment.
unless you want to use `reverse sort`  directly, to avoid the `.rev()` call below
i.e. `items.sort_by(|(_, item_a), (_, item_b)| item_b.elapsed().cmp(&item_a.elapsed()))`

Reply:
yeah, thought of that one, seemed less readable to me
but I'll actually go for it now